### PR TITLE
Add Sonatype OSS Index as advisor

### PR DESCRIPTION
**Pull request by DarthHater**
Hello!

I have added another advisor, and probably done a couple things wrong in the meantime but:

- `SonatypeOssIndex` is available as a new advisor option
- Added a `User-Agent` to both the NexusIQ and OSSIndex client, so that it's easier to tell if this tool is being used (helps us know if we can go support it better!)

I imagine there might be a couple more things to do, like docs, etc... but I figured I'd push code up so people can take a gander.

### CODE REVIEW

File: advisor/src/main/kotlin/advisors/OssIndex.kt [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+        val browseUrl = URL(reference)
+
+        return Vulnerability(displayName, cvssScore, browseUrl)
+    }
+
+    /**
+     * Execute an HTTP request specified by the given [call]. The response status is checked. If everything went
+     * well, the marshalled body of the request is returned; otherwise, the function throws an exception.
+     */
+    private fun <T> execute(call: Call<T>): T {
```
Comment (DarthHater) [inline advisor/src/main/kotlin/advisors/OssIndex.kt:157]: Being this a direct copy of the one from NexusIQ, I'm going to imagine that some Helper, etc... would be a better place for this. Any preference on where?

**Comment by sschuberth** [comment]
Hi @DarthHater, thanks for the contribution! It's a bit unfortunate timing as I just came to the [conclusion](https://github.com/oss-review-toolkit/ort/issues/718#issuecomment-763710308) that it does not make sense for us to integrate OSS Index support due to its severe ToS limitations. Could you maybe provide some arguments why integration into ORT could still make sense (for us, not for you 😉 )?

On the technical side, please take a look at the failing static analysis check (run `./gradlew detekt` locally), and I'll do an initial code review.

File: advisor/build.gradle.kts [hunk 2]
```diff
@@ -26,6 +27,7 @@ plugins {
 
 dependencies {
     api(project(":clients:nexus-iq"))
+    api(project(":clients:oss-index"))
```
Comment (sschuberth) [inline advisor/build.gradle.kts:30]: I'm surprised to see that you've started a separate client project in ORT for this instead of using our very own [ossindex-service-client](https://github.com/sonatype/ossindex-public). Could you please elaborate the details of this decision?

I actually had planned to use *ossindex-public* myself, but ran into some limitations. I started to [address one of them](ossindex-public), but then there were some CLA issues on the Sonetype side. So @DarthHater, maybe you could first help us to resolve these:

https://github.com/sonatype/ossindex-public/issues/7
https://github.com/sonatype/ossindex-public/issues/17
https://github.com/sonatype/ossindex-public/issues/24

File: advisor/src/main/kotlin/advisors/OssIndex.kt (hunk 1 above)
Comment (sschuberth) [reply advisor/src/main/kotlin/advisors/OssIndex.kt:157]: In general the `NexisIq` and `OssIndex` classes have quite some code in common, so maybe it'd make sense to move that common code (incl. this function) to a `SonatypeAdvisor` base class? But the answer also depends on whether we should make use of the *ossindex-public* library instead.

File: clients/nexus-iq/src/main/kotlin/NexusIqService.kt [hunk 3]
```diff
@@ -58,6 +58,7 @@ interface NexusIqService {
                     val requestBuilder = request.newBuilder()
                         .header("X-CSRF-TOKEN", token)
                         .header("Cookie", "CLM-CSRF-TOKEN=$token")
+                        .header("User-Agent", "ort-nexus-iq")
```
Comment (sschuberth) [inline clients/nexus-iq/src/main/kotlin/NexusIqService.kt:61]: As it's unrelated to adding OSS Index as an advisor, please split this out to a separate commit and explain in the commit message what this is good for.

Also, why does the user agent include the service name? I'd rather use a generic string like `.header("User-Agent", "$ORT_NAME/${Environment.ORT_VERSION}")` as we already do in [Downloader.kt](https://github.com/oss-review-toolkit/ort/blob/4be3ab1fbac441ae07fbc8f3f8399582975c0020/downloader/src/main/kotlin/Downloader.kt#L315). I might prepare a global variable for this.

File: reporter-web-app/build.gradle.kts [hunk 4]
```diff
@@ -24,6 +24,7 @@ import org.jetbrains.kotlin.gradle.targets.js.nodejs.NodeJsSetupTask
 import org.jetbrains.kotlin.gradle.targets.js.yarn.YarnPlugin
 import org.jetbrains.kotlin.gradle.targets.js.yarn.YarnSetupTask
 
+
```
Comment (sschuberth) [inline reporter-web-app/build.gradle.kts:27]: Please remove this unnecessary change.

File: settings.gradle.kts [hunk 5]
```diff
@@ -41,6 +42,11 @@ pluginManagement {
             }
         }
     }
+    repositories {
+        maven("https://dl.bintray.com/kotlin/kotlin-eap")
+        mavenCentral()
+        maven("https://plugins.gradle.org/m2/")
+    }
```
Comment (sschuberth) [inline settings.gradle.kts:49]: These shouldn't be required.

**Review by sschuberth — CHANGES_REQUESTED** [review]

File: clients/nexus-iq/src/main/kotlin/NexusIqService.kt (hunk 3 above)
Comment (sschuberth) [reply clients/nexus-iq/src/main/kotlin/NexusIqService.kt:61]: > I might prepare a global variable for this.

See https://github.com/oss-review-toolkit/ort/pull/3531.

**Comment by brianf** [comment]
> Hi @DarthHater, thanks for the contribution! It's a bit unfortunate timing as I just came to the [conclusion](https://github.com/oss-review-toolkit/ort/issues/718#issuecomment-763710308) that it does not make sense for us to integrate OSS Index support due to its severe ToS limitations. Could you maybe provide some arguments why integration into ORT could still make sense (for us, not for you 😉 )?
 
Hi there, we should perhaps have a follow up conversation about the TOS to understand the concerns.

File: advisor/build.gradle.kts (hunk 2 above)
Comment (DarthHater) [reply advisor/build.gradle.kts:30]: Yeah I mostly actually did this out of expedience to learn how the project worked, and Kotlin. I can see if I can help unstick those!

**Comment by DarthHater** [comment]
@brianf is the one to talk to about TOS, and I see he has popped in!

**Comment by sschuberth** [comment]
> Hi there, we should perhaps have a follow up conversation about the TOS to understand the concerns.

Gladly, let's gather at [this issue](https://github.com/oss-review-toolkit/ort/issues/718)!

File: advisor/build.gradle.kts [hunk 6]
```diff
@@ -1,5 +1,6 @@
 /*
  * Copyright (C) 2020 Bosch.IO GmbH
+ * Copyright (C) 2021 Sonatype
```
Comment (sschuberth) [inline advisor/build.gradle.kts:3]: Shouldn't this say "Sonatype Inc." to be more precise?

**Comment by sschuberth** [comment]
@DarthHater, would you have some time to continue working on this? Maybe in a pair-programming session with myself?

**Comment by sschuberth** [comment]
Superseded by https://github.com/oss-review-toolkit/ort/pull/4494.
